packages/e2e-tests/specs/experiments/widgets.test.js

```diff
@@ -0,0 +1,363 @@
+			secondParagraphBlockBoundingBox.y +
+				secondParagraphBlockBoundingBox.height +
+				10
+		);
+
+		// Hover above the last block to trigger the inline inserter between blocks.
+		await page.mouse.move(
+			secondParagraphBlockBoundingBox.x +
+				secondParagraphBlockBoundingBox.width / 2,
+			secondParagraphBlockBoundingBox.y - 10
+		);
+
+		const inserterButton = await page.waitForSelector(
+			'button[aria-label="Add block"][aria-haspopup="true"]'
+		);
+		await inserterButton.click();
+
+		// TODO: The query should be rewritten with role and label.
+		const inserterSearchBox = await page.waitForSelector(
+			'input[type="search"][placeholder="Search for a block"]'
+		);
+		expect(
+			await inserterSearchBox.evaluate(
+				( node ) => node === document.activeElement
+			)
+		).toBe( true );
+
+		await page.keyboard.type( 'Heading' );
+
+		const inserterListBox = await page.$(
+			'[role="listbox"][aria-label="Blocks"]'
+		);
+		const [ headingBlockOption ] = await inserterListBox.$x(
+			'//*[@role="option"][*[text()="Heading"]]'
+		);
+		await headingBlockOption.click();
+
+		// Get the added heading block as second last block.
+		const addedHeadingBlock = await secondParagraphBlock.evaluateHandle(
+			( node ) => node.previousSibling
+		);
+
+		expect(
+			await addedHeadingBlock.evaluate(
+				( node ) => node === document.activeElement
+			)
+		).toBe( true );
+
+		await page.keyboard.type( 'My Heading' );
+
+		const addedHeadingBlockSnapshot = await page.accessibility.snapshot( {
+			root: addedHeadingBlock,
+		} );
+		expect( addedHeadingBlockSnapshot.name ).toBe( 'Block: Heading' );
+		expect( addedHeadingBlockSnapshot.level ).toBe( 2 );
+		expect( addedHeadingBlockSnapshot.value ).toBe( 'My Heading' );
+
+		await saveWidgets();
+		const serializedWidgetAreas = await getSerializedWidgetAreas();
+		expect( serializedWidgetAreas ).toMatchInlineSnapshot( `
+		"<!-- wp:widget-area {\\"id\\":\\"wp_inactive_widgets\\",\\"name\\":\\"Inactive widgets\\"} /-->
+
+		<!-- wp:widget-area {\\"id\\":\\"sidebar-1\\",\\"name\\":\\"Footer #1\\"} -->
+		<!-- wp:paragraph -->
+		<p>First Paragraph</p>
+		<!-- /wp:paragraph -->
+		<!-- wp:heading -->
+		<h2>My Heading</h2>
+		<!-- /wp:heading -->
+		<!-- wp:paragraph -->
+		<p>Second Paragraph</p>
+		<!-- /wp:paragraph -->
+		<!-- /wp:widget-area -->
+
+		<!-- wp:widget-area {\\"id\\":\\"sidebar-2\\",\\"name\\":\\"Footer #2\\"} /-->"
+	` );
+	} );
+} );
+
+async function saveWidgets() {
+	const [ updateButton ] = await page.$x( '//button[text()="Update"]' );
+	await updateButton.click();
+
+	await page.waitForXPath( '//*[text()="Widgets saved."]' );
+
+	// FIXME: The snackbar above is enough for the widget areas to get saved,
+	// but not enough for the widgets to get saved.
+	await page.waitForTimeout( 500 );
+}
+
+async function getSerializedWidgetAreas() {
+	return await page.evaluate( () =>
+		wp.data
+			.select( 'core/edit-widgets' )
+			.getWidgetAreas()
```

Comment:
I wonder if we should assert that `wp.data.select( 'core' ).getWidgets()` matches a snapshot instead so that we're testing the actual conversion of blocks to widgets?

Reply:
Ohh! Didn't know that! Good thing that I pushed this early to get the feedbacks 😅.